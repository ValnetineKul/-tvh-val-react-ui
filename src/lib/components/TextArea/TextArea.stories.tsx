import { createMeta, createTemplate } from '../../story-utils';
import TextArea from './TextArea';

const Template = createTemplate(TextArea);

export const Primitive = Template.bind({});
Primitive.args = {
  label: 'Label',
  rows: 2,
  maxLength: 500,
};
Primitive.storyName = 'TextArea';

export default createMeta({
  component: TextArea,
  title: 'Components/InputFields/TextArea',
  argTypes: {
    onChange: {
      action: { argTypesRegex: '^on.*' },
    },
    inputRef: {
      table: {
        disable: true,
      },
    },
    value: {
      control: {
        disable: true,
      },
    },
  },
});
