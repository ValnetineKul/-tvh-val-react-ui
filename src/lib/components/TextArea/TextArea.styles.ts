import { makeStyles } from '../../../themes/core';

const useStyles = makeStyles({ name: 'TextArea' })((theme) => ({
  root: {
    padding: `${theme.layoutSpacing['spacing/000']} ${theme.layoutSpacing['spacing/300']} ${theme.layoutSpacing['spacing/000']} ${theme.layoutSpacing['spacing/000']} !important`,
  },
  input: {
    padding: `20px ${theme.layoutSpacing['spacing/000']} ${theme.layoutSpacing['spacing/200']} ${theme.layoutSpacing['spacing/300']} !important`,
  },
}));

export default useStyles;
