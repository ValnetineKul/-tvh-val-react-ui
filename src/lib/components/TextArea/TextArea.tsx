import type { FC } from 'react';
import React from 'react';
import type { InputBaseComponentProps } from '@mui/material/InputBase';
import TextField from '../TextField';
import type { DataAttributes } from '../../types/common';
import useStyles from './TextArea.styles';

export interface TextAreaProps {
  label: string;
  id?: string;
  value?: string;
  placeholder?: string;
  errorMessage?: string;
  characterCount?: boolean;
  maxLength?: number;
  rows?: number;
  minRows?: number;
  maxRows?: number;
  disabled?: boolean;
  readOnly?: boolean;
  fullWidth?: boolean;
  required?: boolean;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  inputRef?: React.RefObject<HTMLDivElement>;
  inputElementRef?: React.Ref<HTMLInputElement>;
  autoFocus?: boolean;
  autoSelect?: boolean;
  helpText?: string;
  className?: string;
  inputProps?: InputBaseComponentProps & DataAttributes;
}

const TextArea: FC<TextAreaProps> = ({
  rows = 2,
  maxRows,
  minRows,
  autoFocus = false,
  autoSelect = false,
  inputProps,
  ...props
}) => {
  const { classes } = useStyles();
  const actualRows = maxRows > 0 || minRows > 0 ? undefined : rows;

  return (
    <TextField
      {...props}
      rows={actualRows}
      minRows={minRows}
      maxRows={maxRows}
      multiline
      inputMultilineClassName={classes.root}
      inputClassName={classes.input}
      autoFocus={autoFocus}
      autoSelect={autoSelect}
      inputProps={inputProps}
    />
  );
};

export default TextArea;
