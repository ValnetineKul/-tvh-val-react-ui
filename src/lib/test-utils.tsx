import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
