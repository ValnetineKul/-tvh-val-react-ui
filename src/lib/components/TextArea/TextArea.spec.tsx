import React from 'react';
import { render, userEvent } from '../../test-utils';
import TextArea from './TextArea';

describe('TextArea', () => {
  it('Should render correctly', () => {
    const { container } = render(<TextArea label="label" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should show characters count', () => {
    const { getByText } = render(<TextArea label="label" value="value" maxLength={100} characterCount />);
    expect(getByText('5/100')).toBeInTheDocument();
  });

  it('Should update characters count on value change', () => {
    const { getByText, rerender } = render(<TextArea label="label" value="value" maxLength={100} characterCount />);
    rerender(<TextArea label="label" value="new value" maxLength={100} characterCount />);
    expect(getByText('9/100')).toBeInTheDocument();
  });

  it('Should show error message', () => {
    const { container, getByText } = render(<TextArea label="label" errorMessage="Error message" />);
    expect(getByText('Error message')).toBeInTheDocument();
    expect(container.querySelector('.MuiInputBase-root')?.className).toMatch('Mui-error');
  });

  it('Should trigger onChange', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(<TextArea label="label" value="value" onChange={onChange} />);
    const textarea = getByLabelText('label', { selector: 'textarea', exact: false });
    userEvent.type(textarea, 'new value');
    expect(onChange).toHaveBeenCalledTimes(9);
  });
});
