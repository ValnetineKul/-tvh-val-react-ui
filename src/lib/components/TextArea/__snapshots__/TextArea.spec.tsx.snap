// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Should render correctly 1`] = `
<div
  class="MuiFormControl-root tss-16kerj4-TextField-root mui-1nrlq1o-MuiFormControl-root"
>
  <label
    class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary tss-tsk5sp-InputLabel-root-TextField-label mui-xlr99k-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink="false"
    for="label"
  >
    <span
      class="tss-3-Typography-weightRegular-ref tss-15jpb5e-Typography-root-Typography-weightRegular-InputLabel-label-Typography-body400"
    >
      label
    </span>
  </label>
  <div
    class="tss-bdw984-Surface-root"
  >
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline tss-10-TextField-inputRoot-ref tss-1lxby0w-TextField-inputRoot-TextArea-input mui-mrvkr4-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <textarea
        aria-invalid="false"
        class="tss-1nli869-TextField-input-TextArea-root MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline mui-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
        id="label"
        rows="2"
        style="height: 0px; overflow: hidden;"
      />
      <textarea
        aria-hidden="true"
        class="tss-1nli869-TextField-input-TextArea-root MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline mui-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
        tabindex="-1"
      />
      <fieldset
        aria-hidden="true"
        class="tss-tyiogj-TextField-notchedOutline tss-10-TextField-notchedOutline-ref MuiOutlinedInput-notchedOutline mui-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="mui-nnbavb"
        >
          <span
            class="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    class="tss-19z6szl-TextField-helpers"
  />
</div>
`;
